docs/exts/eliot_metrics.py

```diff
@@ -0,0 +1,109 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, you can obtain one at http://mozilla.org/MPL/2.0/.
+
+"""Generates the documentation for Eliot metrics."""
+
+import importlib
+import sys
+import textwrap
+
+from docutils import nodes
+from docutils.parsers.rst import Directive
+from docutils.statemachine import ViewList
+
+
+def build_table(table):
+    col_size = [0] * len(table[0])
+    for row in table:
+        for i, col in enumerate(row):
+            col_size[i] = max(col_size[i], len(col))
+
+    col_size = [width + 2 for width in col_size]
+
+    yield "  ".join("=" * width for width in col_size)
+    yield "  ".join(
+        header + (" " * (width - len(header)))
+        for header, width in zip(table[0], col_size)
+    )
+    yield "  ".join("=" * width for width in col_size)
+    for row in table[1:]:
+        yield "  ".join(
+            col + (" " * (width - len(col)))
+            for col, width in zip(row, col_size)
+        )
+    yield "  ".join("=" * width for width in col_size)
+
+
+class AutoMetricsDirective(Directive):
+    has_content = False
+    required_arguments = 1
+    optional_arguments = 0
+    final_argument_whitespace = False
+
+    option_spec = {}
+
+    def add_line(self, line, source, *lineno):
+        """Add a line to the result"""
+        self.result.append(line, source, *lineno)
+
+    def generate_docs(self, clspath):
+        modpath, name = clspath.rsplit(".", 1)
+        importlib.import_module(modpath)
+        module = sys.modules[modpath]
+        metrics = getattr(module, name)
+
+        sourcename = "metrics of %s" % clspath
+
+        # First build a table of metric items
+        self.add_line("Table of metrics:", sourcename)
+        self.add_line("", sourcename)
+
+        table = []
+        table.append(("Key", "Type"))
+        for key, metric in metrics.items():
+            table.append((":py:data:`%s`" % key, metric.stat_type))
+
+        for line in build_table(table):
+            self.add_line(line, sourcename)
+
+        self.add_line("", sourcename)
+        self.add_line("Metrics details:", sourcename)
+        self.add_line("", sourcename)
+
+        for key, metric in metrics.items():
+            self.add_line(".. py:data:: %s" % key, sourcename)
```

Comment:
Metrics aren't really "py:data", but I didn't want to write another directive and (ab)using "py:data" gives us what we want: something we can reference, looks ok in HTML, and generates an index entry.